File: training-data/fortran/binary_search.f90
program binary_search
    implicit none
    integer :: sorted(7) = [1, 3, 5, 7, 9, 11, 13]
    print *, search(sorted, 7)
    print *, search(sorted, 4)

contains

    function search(arr, target) result(idx)
        integer, intent(in) :: arr(:)
        integer, intent(in) :: target
        integer :: idx, low, high, mid

        low = 1
        high = size(arr)
        idx = -1
        do while (low <= high)
            mid = (low + high) / 2
            if (arr(mid) == target) then
                idx = mid
                return
            else if (arr(mid) < target) then
                low = mid + 1
            else
                high = mid - 1
            end if
        end do
    end function search

end program binary_search

File: training-data/zig/binary_search.zig
const std = @import("std");

fn binarySearch(arr: []const i32, target: i32) ?usize {
    var low: usize = 0;
    var high: usize = arr.len - 1;
    while (low <= high) {
        const mid = (low + high) / 2;
        if (arr[mid] == target) return mid;
        if (arr[mid] < target) {
            low = mid + 1;
        } else {
            if (mid == 0) return null;
            high = mid - 1;
        }
    }
    return null;
}

pub fn main() void {
    const sorted = [_]i32{ 1, 3, 5, 7, 9, 11, 13 };
    std.debug.print("{any}\n", .{binarySearch(&sorted, 7)});
    std.debug.print("{any}\n", .{binarySearch(&sorted, 4)});
}
